protocols/kad/src/behaviour.rs

```diff
@@ -836,229 +922,232 @@ where
-                    },
-                    QueryInfoInner::GetValue { key, results, .. } => {
-                        let result = match results.len() {
-                            0 => GetValueResult::NotFound {
-                                key,
-                                closest_peers: result.closest_peers.collect()
-                            },
-                            _ => GetValueResult::Found { results },
-                        };
-
-                        let event = KademliaOut::GetValueResult(result);
-
-                        break Async::Ready(NetworkBehaviourAction::GenerateEvent(event));
-                    },
-                    QueryInfoInner::PutValue { key, value } => {
-                        let closest_peers = Vec::from_iter(result.closest_peers);
-                        for peer in &closest_peers {
-                            let event = KademliaHandlerIn::PutValue {
-                                key: key.clone(),
-                                value: value.clone(),
-                                user_data: finished_query,
-                            };
-
-                            if self.connected_peers.contains(peer) {
-                                let event = NetworkBehaviourAction::SendEvent {
-                                    peer_id: peer.clone(),
-                                    event
-                                };
-                                self.queued_events.push(event);
-                            } else {
-                                self.pending_rpcs.push((peer.clone(), event));
-                                self.queued_events.push(NetworkBehaviourAction::DialPeer {
-                                    peer_id: peer.clone(),
-                                });
-                            }
-                        }
-
-                        self.active_writes.insert(finished_query, WriteState::new(key, closest_peers));
-                    },
-                }
-            } else {
-                break Async::NotReady;
+            // No finished query or finished write produced an immediate event.
+            // If no new events have been queued either, signal `NotReady` to
+            // be polled again later.
+            if self.queued_events.is_empty() {
+                return Async::NotReady
             }
         }
     }
 }
 
-/// The result of a `GET_VALUE` query.
-#[derive(Debug, Clone, PartialEq)]
-pub enum GetValueResult {
-    /// The results received from peers. Always contains non-zero number of results.
-    Found { results: Vec<Record> },
-    /// The record wasn't found.
-    NotFound {
-        key: Multihash,
-        closest_peers: Vec<PeerId>
-    }
+/// A quorum w.r.t. the configured replication factor specifies the minimum number of distinct
+/// nodes that must be successfully contacted in order for a query to succeed.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Quorum {
+    One,
+    Two,
```

Comment:
I find `One`, `Two` and `Three` a bit weird, as they are redundant with `N`. Even `Majority` and `All` are redundant, since the user could calculate the actual value themselves.

Reply:
The point is to have a more expressive and convenient (or as Rust folk like to say, _ergonomic_) API for the most-used quorums, and these are at least `One` (for read) and `All` (for write) which yields standard Kademlia semantics. I would be fine with removing `Two` and `Three` but would like to keep `Majority` as well.

> ... since the user could calculate the actual value themselves.

Sure, but it is inconvenient. The quorum should always be evaluated against the configured replication factor (it doesn't have meaningful semantics otherwise), which is done for the user in `put_record` and `get_record`. Instead of having to construct `NonZeroUsize`s, passing a value like `Quorum::One` is surely nicer and communicates more information.